File: core/services/aios-network/network.c
/**
 * AI-OS Network Manager
 * WiFi, Bluetooth, and network management in C.
 * 
 * Compile: gcc -o aios-network network.c -lpthread
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <pthread.h>
#include <errno.h>
#include <dirent.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/ioctl.h>
#include <net/if.h>
#include <arpa/inet.h>

#define SOCKET_PATH "/run/aios/network.sock"

/* ==================== Types ==================== */

typedef struct {
    char name[32];
    char type[16];      /* wifi, ethernet, loopback */
    char state[16];     /* up, down */
    char mac[32];
    char ip[32];
} interface_t;

typedef struct {
    char ssid[64];
    int signal;
    char security[32];
    int connected;
} wifi_network_t;

/* ==================== Globals ==================== */

static volatile int g_running = 1;
static int g_server_fd = -1;

/* ==================== Interface Functions ==================== */

static int get_interfaces(interface_t *ifaces, int max_count) {
    DIR *dir = opendir("/sys/class/net");
    if (!dir) return 0;
    
    int count = 0;
    struct dirent *entry;
    
    while ((entry = readdir(dir)) && count < max_count) {
        if (entry->d_name[0] == '.') continue;
        
        interface_t *iface = &ifaces[count];
        strncpy(iface->name, entry->d_name, sizeof(iface->name) - 1);
        
        /* Get type */
        char path[256];
        snprintf(path, sizeof(path), "/sys/class/net/%s/wireless", entry->d_name);
        if (access(path, F_OK) == 0) {
            strcpy(iface->type, "wifi");
        } else if (strcmp(entry->d_name, "lo") == 0) {
            strcpy(iface->type, "loopback");
        } else if (strncmp(entry->d_name, "eth", 3) == 0 || 
                   strncmp(entry->d_name, "en", 2) == 0) {
            strcpy(iface->type, "ethernet");
        } else {
            strcpy(iface->type, "unknown");
        }
        
        /* Get state */
        snprintf(path, sizeof(path), "/sys/class/net/%s/operstate", entry->d_name);
        FILE *f = fopen(path, "r");
        if (f) {
            fgets(iface->state, sizeof(iface->state), f);
            iface->state[strcspn(iface->state, "\n")] = '\0';
            fclose(f);
        }
        
        /* Get MAC */
        snprintf(path, sizeof(path), "/sys/class/net/%s/address", entry->d_name);
        f = fopen(path, "r");
        if (f) {
            fgets(iface->mac, sizeof(iface->mac), f);
            iface->mac[strcspn(iface->mac, "\n")] = '\0';
            fclose(f);
        }
        
        /* Get IP via ioctl */
        int sock = socket(AF_INET, SOCK_DGRAM, 0);
        if (sock >= 0) {
            struct ifreq ifr;
            strncpy(ifr.ifr_name, iface->name, IFNAMSIZ - 1);
            if (ioctl(sock, SIOCGIFADDR, &ifr) == 0) {
                struct sockaddr_in *addr = (struct sockaddr_in *)&ifr.ifr_addr;
                inet_ntop(AF_INET, &addr->sin_addr, iface->ip, sizeof(iface->ip));
            }
            close(sock);
        }
        
        count++;
    }
    
    closedir(dir);
    return count;
}

/* ==================== WiFi Functions ==================== */

static int wifi_enabled(void) {
    FILE *fp = popen("nmcli radio wifi 2>/dev/null", "r");
    if (!fp) return -1;
    
    char buf[32];
    fgets(buf, sizeof(buf), fp);
    pclose(fp);
    
    return strstr(buf, "enabled") ? 1 : 0;
}

static int wifi_set_enabled(int enabled) {
    char cmd[64];
    snprintf(cmd, sizeof(cmd), "nmcli radio wifi %s >/dev/null 2>&1", 
             enabled ? "on" : "off");
    return system(cmd) == 0 ? 0 : -1;
}

static int wifi_scan(wifi_network_t *networks, int max_count) {
    /* Trigger scan */
    system("nmcli device wifi rescan 2>/dev/null");
    usleep(500000);
    
    FILE *fp = popen("nmcli -t -f ACTIVE,SSID,SIGNAL,SECURITY device wifi list 2>/dev/null", "r");
    if (!fp) return 0;
    
    int count = 0;
    char line[256];
    
    while (fgets(line, sizeof(line), fp) && count < max_count) {
        line[strcspn(line, "\n")] = '\0';
        
        char *active = strtok(line, ":");
        char *ssid = strtok(NULL, ":");
        char *signal = strtok(NULL, ":");
        char *security = strtok(NULL, ":");
        
        if (!ssid || !ssid[0]) continue;
        
        wifi_network_t *net = &networks[count];
        strncpy(net->ssid, ssid, sizeof(net->ssid) - 1);
        net->signal = signal ? atoi(signal) : 0;
        strncpy(net->security, security ? security : "Open", sizeof(net->security) - 1);
        net->connected = (active && strcmp(active, "yes") == 0);
        
        count++;
    }
    
    pclose(fp);
    return count;
}

static int wifi_connect(const char *ssid, const char *password) {
    char cmd[512];
    if (password && password[0]) {
        snprintf(cmd, sizeof(cmd), 
            "nmcli device wifi connect \"%s\" password \"%s\" >/dev/null 2>&1",
            ssid, password);
    } else {
        snprintf(cmd, sizeof(cmd), 
            "nmcli device wifi connect \"%s\" >/dev/null 2>&1", ssid);
    }
    return system(cmd) == 0 ? 0 : -1;
}

static int wifi_disconnect(void) {
    return system("nmcli device disconnect wlan0 >/dev/null 2>&1") == 0 ? 0 : -1;
}

/* ==================== Bluetooth Functions ==================== */

static int bluetooth_enabled(void) {
    FILE *fp = popen("bluetoothctl show 2>/dev/null | grep 'Powered: yes'", "r");
    if (!fp) return -1;
    
    char buf[64];
    int enabled = (fgets(buf, sizeof(buf), fp) != NULL);
    pclose(fp);
    
    return enabled;
}

static int bluetooth_set_enabled(int enabled) {
    char cmd[64];
    snprintf(cmd, sizeof(cmd), "bluetoothctl power %s >/dev/null 2>&1",
             enabled ? "on" : "off");
    return system(cmd) == 0 ? 0 : -1;
}

/* ==================== IPC Server ==================== */

static void handle_request(int client_fd, const char *msg) {
    char response[8192] = "{\"status\":\"ok\"}";
    
    if (strstr(msg, "\"cmd\":\"interfaces\"")) {
        interface_t ifaces[16];
        int count = get_interfaces(ifaces, 16);
        
        char *ptr = response;
        ptr += sprintf(ptr, "{\"status\":\"ok\",\"interfaces\":[");
        
        for (int i = 0; i < count; i++) {
            if (i > 0) ptr += sprintf(ptr, ",");
            ptr += sprintf(ptr, 
                "{\"name\":\"%s\",\"type\":\"%s\",\"state\":\"%s\",\"mac\":\"%s\",\"ip\":\"%s\"}",
                ifaces[i].name, ifaces[i].type, ifaces[i].state, 
                ifaces[i].mac, ifaces[i].ip);
        }
        ptr += sprintf(ptr, "]}");
    }
    else if (strstr(msg, "\"cmd\":\"wifi_status\"")) {
        int enabled = wifi_enabled();
        snprintf(response, sizeof(response), 
            "{\"status\":\"ok\",\"wifi_enabled\":%s}", 
            enabled ? "true" : "false");
    }
    else if (strstr(msg, "\"cmd\":\"wifi_enable\"")) {
        int enable = strstr(msg, "\"enable\":true") != NULL;
        wifi_set_enabled(enable);
        snprintf(response, sizeof(response), 
            "{\"status\":\"ok\",\"wifi_enabled\":%s}", enable ? "true" : "false");
    }
    else if (strstr(msg, "\"cmd\":\"wifi_scan\"")) {
        wifi_network_t networks[32];
        int count = wifi_scan(networks, 32);
        
        char *ptr = response;
        ptr += sprintf(ptr, "{\"status\":\"ok\",\"networks\":[");
        
        for (int i = 0; i < count; i++) {
            if (i > 0) ptr += sprintf(ptr, ",");
            ptr += sprintf(ptr, 
                "{\"ssid\":\"%s\",\"signal\":%d,\"security\":\"%s\",\"connected\":%s}",
                networks[i].ssid, networks[i].signal, networks[i].security,
                networks[i].connected ? "true" : "false");
        }
        ptr += sprintf(ptr, "]}");
    }
    else if (strstr(msg, "\"cmd\":\"wifi_connect\"")) {
        char ssid[64] = "", password[128] = "";
        char *p;
        
        if ((p = strstr(msg, "\"ssid\":\""))) {
            sscanf(p, "\"ssid\":\"%63[^\"]\"", ssid);
        }
        if ((p = strstr(msg, "\"password\":\""))) {
            sscanf(p, "\"password\":\"%127[^\"]\"", password);
        }
        
        int result = wifi_connect(ssid, password);
        snprintf(response, sizeof(response), 
            "{\"status\":\"%s\",\"message\":\"%s\"}", 
            result == 0 ? "ok" : "error",
            result == 0 ? "Connected" : "Connection failed");
    }
    else if (strstr(msg, "\"cmd\":\"wifi_disconnect\"")) {
        wifi_disconnect();
    }
    else if (strstr(msg, "\"cmd\":\"bluetooth_status\"")) {
        int enabled = bluetooth_enabled();
        snprintf(response, sizeof(response), 
            "{\"status\":\"ok\",\"bluetooth_enabled\":%s}", 
            enabled ? "true" : "false");
    }
    else if (strstr(msg, "\"cmd\":\"bluetooth_enable\"")) {
        int enable = strstr(msg, "\"enable\":true") != NULL;
        bluetooth_set_enabled(enable);
    }
    
    /* Send response */
    uint32_t len = htonl(strlen(response));
    send(client_fd, &len, 4, 0);
    send(client_fd, response, strlen(response), 0);
}

static void *client_handler(void *arg) {
    int client_fd = *(int *)arg;
    free(arg);
    
    while (1) {
        uint32_t length;
        if (recv(client_fd, &length, 4, 0) != 4) break;
        length = ntohl(length);
        
        if (length > 4096) break;
        
        char *msg = malloc(length + 1);
        if (!msg) break;
        
        if (recv(client_fd, msg, length, 0) != length) {
            free(msg);
            break;
        }
        msg[length] = '\0';
        
        handle_request(client_fd, msg);
        free(msg);
    }
    
    close(client_fd);
    return NULL;
}

static int run_server(void) {
    g_server_fd = socket(AF_UNIX, SOCK_STREAM, 0);
    if (g_server_fd < 0) {
        perror("socket");
        return -1;
    }
    
    struct sockaddr_un addr = {0};
    addr.sun_family = AF_UNIX;
    strncpy(addr.sun_path, SOCKET_PATH, sizeof(addr.sun_path) - 1);
    
    unlink(SOCKET_PATH);
    
    if (bind(g_server_fd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
        perror("bind");
        return -1;
    }
    
    chmod(SOCKET_PATH, 0666);
    listen(g_server_fd, 5);
    
    printf("[NETWORK] Listening on %s\n", SOCKET_PATH);
    
    while (g_running) {
        int *client = malloc(sizeof(int));
        *client = accept(g_server_fd, NULL, NULL);
        
        if (*client < 0) {
            free(client);
            continue;
        }
        
        pthread_t thread;
        pthread_create(&thread, NULL, client_handler, client);
        pthread_detach(thread);
    }
    
    return 0;
}

/* ==================== Main ==================== */

static void signal_handler(int sig) {
    g_running = 0;
}

int main(int argc, char *argv[]) {
    printf("[NETWORK] AI-OS Network Manager starting...\n");
    
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);
    
    mkdir("/run/aios", 0755);
    
    run_server();
    
    if (g_server_fd >= 0) {
        close(g_server_fd);
        unlink(SOCKET_PATH);
    }
    
    printf("[NETWORK] Network manager stopped\n");
    return 0;
}

File: core/services/aios-network/network.py
#!/usr/bin/env python3
"""
AI-OS Network Manager
Handles WiFi, Ethernet, Bluetooth, and VPN configuration.
"""

import os
import subprocess
import json
from dataclasses import dataclass
from typing import List, Dict, Optional, Any
from pathlib import Path
import logging

logger = logging.getLogger('aios-network')


@dataclass
class NetworkInterface:
    """Network interface information"""
    name: str
    type: str  # ethernet, wifi, loopback
    state: str  # up, down, dormant
    mac_address: str = ""
    ip_address: str = ""
    gateway: str = ""
    dns: List[str] = None


@dataclass
class WifiNetwork:
    """WiFi network information"""
    ssid: str
    signal: int
    security: str
    connected: bool = False
    saved: bool = False


@dataclass
class BluetoothDevice:
    """Bluetooth device information"""
    address: str
    name: str
    paired: bool
    connected: bool
    device_type: str = "unknown"


class NetworkManager:
    """AI-OS Network Manager"""
    
    def __init__(self):
        self._check_tools()
    
    def _check_tools(self):
        """Check for required tools"""
        self.has_nmcli = self._cmd_exists('nmcli')
        self.has_iw = self._cmd_exists('iw')
        self.has_bluetoothctl = self._cmd_exists('bluetoothctl')
    
    def _cmd_exists(self, cmd: str) -> bool:
        """Check if a command exists"""
        return subprocess.run(['which', cmd], capture_output=True).returncode == 0
    
    def _run(self, cmd: List[str], timeout: int = 10) -> Optional[str]:
        """Run a command and return output"""
        try:
            result = subprocess.run(cmd, capture_output=True, text=True, timeout=timeout)
            return result.stdout
        except Exception as e:
            logger.error(f"Command failed: {' '.join(cmd)}: {e}")
            return None
    
    # ==================== General ====================
    
    def get_interfaces(self) -> List[NetworkInterface]:
        """Get all network interfaces"""
        interfaces = []
        
        try:
            for iface in Path('/sys/class/net').iterdir():
                name = iface.name
                
                # Get type
                if (iface / 'wireless').exists():
                    itype = 'wifi'
                elif name.startswith('eth') or name.startswith('en'):
                    itype = 'ethernet'
                elif name == 'lo':
                    itype = 'loopback'
                else:
                    itype = 'unknown'
                
                # Get state
                state = (iface / 'operstate').read_text().strip()
                
                # Get MAC
                mac = (iface / 'address').read_text().strip()
                
                interfaces.append(NetworkInterface(
                    name=name,
                    type=itype,
                    state=state,
                    mac_address=mac
                ))
        except Exception as e:
            logger.error(f"Failed to get interfaces: {e}")
        
        return interfaces
    
    def get_connection_status(self) -> Dict[str, Any]:
        """Get overall connection status"""
        if self.has_nmcli:
            output = self._run(['nmcli', '-t', '-f', 'TYPE,STATE,CONNECTION', 'device'])
            if output:
                for line in output.strip().split('\n'):
                    parts = line.split(':')
                    if len(parts) >= 3 and parts[1] == 'connected':
                        return {
                            'connected': True,
                            'type': parts[0],
                            'connection': parts[2]
                        }
        
        return {'connected': False}
    
    # ==================== WiFi ====================
    
    def scan_wifi(self) -> List[WifiNetwork]:
        """Scan for WiFi networks"""
        networks = []
        
        if self.has_nmcli:
            # Trigger rescan
            self._run(['nmcli', 'device', 'wifi', 'rescan'])
            
            output = self._run(['nmcli', '-t', '-f', 'ACTIVE,SSID,SIGNAL,SECURITY', 'device', 'wifi', 'list'])
            if output:
                for line in output.strip().split('\n'):
                    parts = line.split(':')
                    if len(parts) >= 4 and parts[1]:  # Has SSID
                        networks.append(WifiNetwork(
                            ssid=parts[1],
                            signal=int(parts[2]) if parts[2].isdigit() else 0,
                            security=parts[3],
                            connected=parts[0] == 'yes'
                        ))
        
        return networks
    
    def connect_wifi(self, ssid: str, password: Optional[str] = None) -> bool:
        """Connect to a WiFi network"""
        if not self.has_nmcli:
            return False
        
        cmd = ['nmcli', 'device', 'wifi', 'connect', ssid]
        if password:
            cmd.extend(['password', password])
        
        result = subprocess.run(cmd, capture_output=True)
        return result.returncode == 0
    
    def disconnect_wifi(self) -> bool:
        """Disconnect from current WiFi"""
        if not self.has_nmcli:
            return False
        
        result = subprocess.run(['nmcli', 'device', 'disconnect', 'wlan0'], capture_output=True)
        return result.returncode == 0
    
    def wifi_enabled(self) -> bool:
        """Check if WiFi is enabled"""
        if self.has_nmcli:
            output = self._run(['nmcli', 'radio', 'wifi'])
            return output and 'enabled' in output.lower()
        return False
    
    def set_wifi_enabled(self, enabled: bool) -> bool:
        """Enable/disable WiFi"""
        if not self.has_nmcli:
            return False
        
        state = 'on' if enabled else 'off'
        result = subprocess.run(['nmcli', 'radio', 'wifi', state], capture_output=True)
        return result.returncode == 0
    
    def get_saved_networks(self) -> List[str]:
        """Get list of saved WiFi networks"""
        if not self.has_nmcli:
            return []
        
        output = self._run(['nmcli', '-t', '-f', 'NAME,TYPE', 'connection'])
        if output:
            return [
                line.split(':')[0]
                for line in output.strip().split('\n')
                if ':802-11-wireless' in line
            ]
        return []
    
    def forget_network(self, name: str) -> bool:
        """Forget a saved network"""
        if not self.has_nmcli:
            return False
        
        result = subprocess.run(['nmcli', 'connection', 'delete', name], capture_output=True)
        return result.returncode == 0
    
    # ==================== Bluetooth ====================
    
    def scan_bluetooth(self, duration: int = 5) -> List[BluetoothDevice]:
        """Scan for Bluetooth devices"""
        devices = []
        
        if not self.has_bluetoothctl:
            return devices
        
        # Start scan
        subprocess.run(['bluetoothctl', 'scan', 'on'], capture_output=True, timeout=2)
        import time
        time.sleep(duration)
        subprocess.run(['bluetoothctl', 'scan', 'off'], capture_output=True, timeout=2)
        
        # Get devices
        output = self._run(['bluetoothctl', 'devices'])
        if output:
            for line in output.strip().split('\n'):
                if line.startswith('Device'):
                    parts = line.split(' ', 2)
                    if len(parts) >= 3:
                        devices.append(BluetoothDevice(
                            address=parts[1],
                            name=parts[2],
                            paired=False,
                            connected=False
                        ))
        
        return devices
    
    def get_paired_devices(self) -> List[BluetoothDevice]:
        """Get paired Bluetooth devices"""
        devices = []
        
        if not self.has_bluetoothctl:
            return devices
        
        output = self._run(['bluetoothctl', 'paired-devices'])
        if output:
            for line in output.strip().split('\n'):
                if line.startswith('Device'):
                    parts = line.split(' ', 2)
                    if len(parts) >= 3:
                        devices.append(BluetoothDevice(
                            address=parts[1],
                            name=parts[2],
                            paired=True,
                            connected=False  # Would need to check
                        ))
        
        return devices
    
    def connect_bluetooth(self, address: str) -> bool:
        """Connect to a Bluetooth device"""
        if not self.has_bluetoothctl:
            return False
        
        result = subprocess.run(['bluetoothctl', 'connect', address], capture_output=True, timeout=10)
        return result.returncode == 0
    
    def disconnect_bluetooth(self, address: str) -> bool:
        """Disconnect from a Bluetooth device"""
        if not self.has_bluetoothctl:
            return False
        
        result = subprocess.run(['bluetoothctl', 'disconnect', address], capture_output=True)
        return result.returncode == 0
    
    def pair_bluetooth(self, address: str) -> bool:
        """Pair with a Bluetooth device"""
        if not self.has_bluetoothctl:
            return False
        
        # Trust first
        subprocess.run(['bluetoothctl', 'trust', address], capture_output=True)
        
        result = subprocess.run(['bluetoothctl', 'pair', address], capture_output=True, timeout=30)
        return result.returncode == 0
    
    def bluetooth_enabled(self) -> bool:
        """Check if Bluetooth is enabled"""
        if not self.has_bluetoothctl:
            return False
        
        output = self._run(['bluetoothctl', 'show'])
        return output and 'Powered: yes' in output
    
    def set_bluetooth_enabled(self, enabled: bool) -> bool:
        """Enable/disable Bluetooth"""
        if not self.has_bluetoothctl:
            return False
        
        state = 'on' if enabled else 'off'
        result = subprocess.run(['bluetoothctl', 'power', state], capture_output=True)
        return result.returncode == 0
    
    # ==================== VPN ====================
    
    def get_vpn_connections(self) -> List[str]:
        """Get available VPN connections"""
        if not self.has_nmcli:
            return []
        
        output = self._run(['nmcli', '-t', '-f', 'NAME,TYPE', 'connection'])
        if output:
            return [
                line.split(':')[0]
                for line in output.strip().split('\n')
                if ':vpn' in line
            ]
        return []
    
    def connect_vpn(self, name: str) -> bool:
        """Connect to a VPN"""
        if not self.has_nmcli:
            return False
        
        result = subprocess.run(['nmcli', 'connection', 'up', name], capture_output=True)
        return result.returncode == 0
    
    def disconnect_vpn(self, name: str) -> bool:
        """Disconnect from a VPN"""
        if not self.has_nmcli:
            return False
        
        result = subprocess.run(['nmcli', 'connection', 'down', name], capture_output=True)
        return result.returncode == 0
    
    # ==================== Hotspot ====================
    
    def create_hotspot(self, ssid: str, password: str) -> bool:
        """Create a WiFi hotspot"""
        if not self.has_nmcli:
            return False
        
        result = subprocess.run([
            'nmcli', 'device', 'wifi', 'hotspot',
            'ssid', ssid,
            'password', password
        ], capture_output=True)
        return result.returncode == 0
    
    def stop_hotspot(self) -> bool:
        """Stop the WiFi hotspot"""
        if not self.has_nmcli:
            return False
        
        result = subprocess.run(['nmcli', 'connection', 'down', 'Hotspot'], capture_output=True)
        return result.returncode == 0


# Example usage
if __name__ == '__main__':
    nm = NetworkManager()
    
    print("Network Interfaces:")
    for iface in nm.get_interfaces():
        print(f"  {iface.name}: {iface.type} ({iface.state})")
    
    print("\nConnection Status:", nm.get_connection_status())
    
    print("\nWiFi Networks:")
    for net in nm.scan_wifi():
        status = "●" if net.connected else "○"
        print(f"  {status} {net.ssid} ({net.signal}%) [{net.security}]")
    
    print("\nBluetooth Paired Devices:")
    for dev in nm.get_paired_devices():
        print(f"  {dev.name} ({dev.address})")
